### PR TITLE
[LowerSwitch] Don't let pass manager handle the dependency

### PR DESCRIPTION
**Pull request by ruiling**
Some passes has limitation that only support simple terminators: branch/unreachable/return. Right now, they ask the pass manager to add LowerSwitch pass to eliminate `switch`. Let's manage such kind of pass dependency by ourselves. Also add the assertion in the related passes.

### CODE REVIEW

File: llvm/test/CodeGen/AMDGPU/llc-pipeline.ll [hunk 1]
```diff
@@ -60,8 +60,6 @@
 ; GCN-O0-NEXT:        Cycle Info Analysis
 ; GCN-O0-NEXT:        Uniformity Analysis
 ; GCN-O0-NEXT:        Unify divergent function exit nodes
-; GCN-O0-NEXT:        Lazy Value Information Analysis
-; GCN-O0-NEXT:        Lower SwitchInst's to branches
```
Comment (jayfoad) [inline llvm/test/CodeGen/AMDGPU/llc-pipeline.ll:64]: Removing this is nice. I prototyped an alternative way to do it: https://github.com/jayfoad/llvm-project/commit/c03cbd4d2f584d24e2100fee529f2ba1d18c345d

See also #59830

File: llvm/lib/Transforms/Utils/BasicBlockUtils.cpp [hunk 2]
```diff
@@ -2076,3 +2076,11 @@ void llvm::InvertBranch(BranchInst *PBI, IRBuilderBase &Builder) {
   PBI->setCondition(NewCond);
   PBI->swapSuccessors();
 }
+
+bool llvm::hasOnlySimpleTerminator(const BasicBlock *BB) {
+  auto *Term = BB->getTerminator();
+  if (isa<ReturnInst>(Term) || isa<UnreachableInst>(Term) ||
+      isa<BranchInst>(Term))
+    return true;
+  return false;
```
Comment (jayfoad) [inline llvm/lib/Transforms/Utils/BasicBlockUtils.cpp:2085]:
```suggestion
  return isa<ReturnInst>(Term) || isa<UnreachableInst>(Term) ||
      isa<BranchInst>(Term);
```

**Comment by jayfoad** [comment]
> Some passes has limitation that only support simple terminators: branch/unreachable/return. Right now, they ask the pass manager to add LowerSwitch pass to eliminate `switch`. Let's manage such kind of pass dependency by ourselves. Also add the assertion in the related passes.

I have no objections, but why do you think it is better to do it this way?

**Comment by ruiling** [comment]
> > Some passes has limitation that only support simple terminators: branch/unreachable/return. Right now, they ask the pass manager to add LowerSwitch pass to eliminate `switch`. Let's manage such kind of pass dependency by ourselves. Also add the assertion in the related passes.
> 
> I have no objections, but why do you think it is better to do it this way?

In https://reviews.llvm.org/D89026, @aeubanks said "The NPM doesn't support adding a non-analysis pass as a dependency of another, so I had to add -lowerswitch to some tests or pin them to the legacy PM." I guess it is a burden to support such kind of non-analysis pass dependency in NPM and it is not quite useful in general.

**Comment by aeubanks** [comment]
I'm not really seeing the value in doing this unless we go toward completely removing the whole "required pass" concept of the legacy PM. But the legacy PM treats dependent analyses and passes the same way, so you'd have to rewrite the legacy PM and you might as well port everything to the new PM instead.

**Comment by ruiling** [comment]
> I'm not really seeing the value in doing this unless we go toward completely removing the whole "required pass" concept of the legacy PM. But the legacy PM treats dependent analyses and passes the same way, so you'd have to rewrite the legacy PM and you might as well port everything to the new PM instead.

By saying "doing this", do you mean the change here? The change here will help remove one extra run of this pass for AMDGPU and also make the tests for these passes can run under NPM. I think we want to switch to NPM anyway, right?

**Comment by jayfoad** [comment]
> But the legacy PM treats dependent analyses and passes the same way, so you'd have to rewrite the legacy PM

That's obviously not gonna happen. But just because the legacy PM _supports_ using a pass like an analysis, that does not mean we have _use_ it like that in this case.

> The change here will help remove one extra run of this pass for AMDGPU

That is a clear win. The question is: do we want to do it like in this patch? Or by declaring that more passes preserve LowerSwitch as in https://github.com/jayfoad/llvm-project/commit/c03cbd4d2f584d24e2100fee529f2ba1d18c345d.

**Comment by ruiling** [comment]
> That is a clear win. The question is: do we want to do it like in this patch? Or by declaring that more passes preserve LowerSwitch as in [jayfoad@c03cbd4](https://github.com/jayfoad/llvm-project/commit/c03cbd4d2f584d24e2100fee529f2ba1d18c345d).

If I understand the situation correctly, your change only work for legacy pass manager? As our goal is to switch to New PM, I think it is better to just remove the `addPreservedID(LowerSwitchID)` and `addRequiredID(LowerSwitchID)` from the passes which depend on `LowerSwitch`? Because that would be what we want the code looks like under NewPM.

**Comment by jayfoad** [comment]
> your change only work for legacy pass manager?

Yes. The new pass manager does not support codegen passes.

**Review by aeubanks — COMMENTED** [review]
I didn't understand that this saves a pass run, this lgtm then

(also I thought I already sent out these comments, sorry for the delay)

File: llvm/lib/Target/AMDGPU/AMDGPUUnifyDivergentExitNodes.cpp [hunk 3]
```diff
@@ -192,6 +191,9 @@ BasicBlock *AMDGPUUnifyDivergentExitNodesImpl::unifyReturnBlockSet(
 bool AMDGPUUnifyDivergentExitNodesImpl::run(Function &F, DominatorTree *DT,
                                             const PostDominatorTree &PDT,
                                             const UniformityInfo &UA) {
+  for (auto &BB : F)
+    assert(hasOnlySimpleTerminator(&BB));
```
[the remark this answers is not in the document]
Comment (ruiling) [reply llvm/lib/Target/AMDGPU/AMDGPUUnifyDivergentExitNodes.cpp]: Thanks for the suggestion, but I think there might be also other unsupported terminator goes here, so I just add a message "unsupported block terminator".

**Review by aeubanks — APPROVED** [review]

**Pull request merged**